### PR TITLE
sql: tighten spans at the beginning and end to not scan interleaved rows

### PR DESCRIPTION
**Pull request by richardwu**
I apologize in advance for the really long PR (80% of it are tests, I promise!).

Still to do (probably in a later PR so not to over-bloat this one) is to do the same during/after `partitionSpans` in DistSQL.

### CODE REVIEW

**Comment by RaduBerinde** [comment]
Review status: 0 of 25 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/aggregate, line 719 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-Kzk7mFAgM7nAkzA9WI5:-Kzk7mFAgM7nAkzA9WI6:b-k3lhg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/aggregate#L719)):*
> ```
> 2
> 
> # The span looks off because NULLASC (the first '#') and
> ```

Ha, I was just about to ask if this is the case. This makes it hard to read these tests, it looks like an empty span. I expect this to cause confusion when debugging things. We should disambiguate, looks like a fairly simple change in `encoding.go:prettyPrintFirstValue`. Perhaps `!NULL` instead of the first`#`

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-KzkBHPmnDMg9AZvOkwm:-KzkBHPmnDMg9AZvOkwn:b4ab4r8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L1)):*
> ```
> # LogicTest: 5node-distsql
> ```

wow, these tests are quite comprehensive

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-KzkIjAZbAy2Wd6U1Pn_:-KzkIjAZbAy2Wd6U1Pna:b-4a1oku) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L27)):*
> ```
>   a INT PRIMARY KEY
> ) INTERLEAVE IN PARENT t (a)
> ```

maybe add some tests with a table which has an index that is interleaved (e.g. `CREATE INDEX p1_id ON p1_1 (i, d) INTERLEAVE IN PARENT`)

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-Kzk9wHN8xHsv9BRV8xK:-Kzk9wHN8xHsv9BRV8xL:b-ayxut1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L182)):*
> ```
> SELECT "URL" FROM [EXPLAIN (DISTSQL) SELECT * FROM t WHERE a <= 4]
> ----
> https://cockroachdb.github.io/distsqlplan/decode.html?eJzMkTFr9DAMhvfvV3y8swpx0slT1lvuytGtZHBjcQRyUZAdaDny30vs4ZojLS0d2tGSHj2ydMEgnvfuzAH2CQaEEoQKDWFUaTkE0SWVC3f-BbYgdMM4xSXcEFpRhr0gdrFnWDy6556P7DwrCJ6j6_rUfNTu7PS1jiAcpmj_14bqkuoKzUyQKV57huhODGtm-sB71U2DqGdlv5I188Zke7mT8aZsW1yuxOaXPlz-gUVveI8cRhkCf2mTxXII9ifOVwsyacsPKm3S5OchcSngOcScNfmxG1IqDfgeNp_C9yu4uIXLn5irb8HN_O8tAAD__8RJHKE=
> ```

What's the point of including these plans? They don't show the spans..

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-KzkC8CdlwcJoh6Bgwqw:-KzkC8Ceb-9JB_u7Gvgg:b3l7w3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L698)):*
> ```
> # In some cases, we cannot tighten the span if the subsequent column is
> # in DESC order and is NULLable.
> # If we had tightened this to /1-/1/#, then /1/NULL would be omitted.
> ```

This special case looks strange to me. I understand the problem, but I believe that if there are more key columns, we can't have an interleaving at this position anyway. In fact, I don't think we can have an interleaving where the parent is an index at all (so the parent is always a primary index, and PK columns are not-nullable). 

We should probably handle the cases where there can't be an interleaving the same way, regardless if we're dealing with a desc/nullable index.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235)*

**Comment by richardwu** [comment]
Review status: 0 of 25 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/aggregate, line 719 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-Kzk7mFAgM7nAkzA9WI5:-L-2LrcyEBpAFRgc7vc4:bcv3mro) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/aggregate#L719)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ha, I was just about to ask if this is the case. This makes it hard to read these tests, it looks like an empty span. I expect this to cause confusion when debugging things. We should disambiguate, looks like a fairly simple change in `encoding.go:prettyPrintFirstValue`. Perhaps `!NULL` instead of the first`#`
</blockquote></details>

It's rather unfortunate there's no clean way to annotate whether the `#` is used as an interleaved sentinel or a descending not null token in the pretty printer. While pretty spans for descending indexes are somewhat wonky ATM (see #20203), we'd have to reconcile that a query with `IS NOT NULL` on a descending index will produce a different pretty span than an ascending index (with the `!NULL` disambiguation).

For example, the query on an ascending index
```
EXPLAIN SELECT * FROM t@asc_idx WHERE a IS NOT NULL
```
will now produce the pretty span
```
/!NULL-
```
whereas the query on a descending index
```
EXPLAIN SELECT * FROM t@desc_idx WHERE a IS NOT NULL
```
will produce the pretty span
```
-/#
```

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-KzkIjAZbAy2Wd6U1Pn_:-L-2VkyLMd3nXR3JLe9s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L27)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

maybe add some tests with a table which has an index that is interleaved (e.g. `CREATE INDEX p1_id ON p1_1 (i, d) INTERLEAVE IN PARENT`)
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 182 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-Kzk9wHN8xHsv9BRV8xK:-L-2X0ulOgUgQU-X2zM_:b-598vun) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L182)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What's the point of including these plans? They don't show the spans..
</blockquote></details>

They were included preemptively to ensure that once we start tightening spans after distsql partitioning, we'd still be reading from the appropriate nodes.

E.g. the table is split at the key `/2/#/52/1`, so an overall span like
```
/1 - /3/#
`` 
would be partitioned into
```
/1 - /2/#/52/1 (node 1)
/2/#/52/1 - /3/# (node 2)
```
and then tightened to
```
/1 - /3
/3 - /3/#
```
which should still be scanned on their respective nodes.

Now that I think about it, using the distsql plan to validate this might be too implementation-focused rather than correctness-focused. I believe the query results would be sufficient to test for correctness of tightening the spans after partitioning.
```


---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-KzkC8CdlwcJoh6Bgwqw:-L-2WZfwR8iPmz9iW_3D:beb3pm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L698)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This special case looks strange to me. I understand the problem, but I believe that if there are more key columns, we can't have an interleaving at this position anyway. In fact, I don't think we can have an interleaving where the parent is an index at all (so the parent is always a primary index, and PK columns are not-nullable). 

We should probably handle the cases where there can't be an interleaving the same way, regardless if we're dealing with a desc/nullable index.
</blockquote></details>

Just to clarify: do you mean we should just shortcut any tightening the moment we see that the index is not the primary index?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 25 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/aggregate, line 719 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-Kzk7mFAgM7nAkzA9WI5:-L-2_oDKUs19HcplMaLo:b-ylalmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/aggregate#L719)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

It's rather unfortunate there's no clean way to annotate whether the `#` is used as an interleaved sentinel or a descending not null token in the pretty printer. While pretty spans for descending indexes are somewhat wonky ATM (see #20203), we'd have to reconcile that a query with `IS NOT NULL` on a descending index will produce a different pretty span than an ascending index (with the `!NULL` disambiguation).

For example, the query on an ascending index
```
EXPLAIN SELECT * FROM t@asc_idx WHERE a IS NOT NULL
```
will now produce the pretty span
```
/!NULL-
```
whereas the query on a descending index
```
EXPLAIN SELECT * FROM t@desc_idx WHERE a IS NOT NULL
```
will produce the pretty span
```
-/#
```
</blockquote></details>

I'm thinking we can pass the direction info to prettyPrint. That might be helpful for #20203 too

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-KzkC8CdlwcJoh6Bgwqw:-L-2aNtpOHfJfXxzoBU6:b-3min0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L698)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

Just to clarify: do you mean we should just shortcut any tightening the moment we see that the index is not the primary index?
</blockquote></details>

Yes, and also if the spans don't cover the entire primary index. If we have primary index on `a,b`, I believe we can't have interleaved child tables which share just `a` (right?). In this case, spans like `/1` shouldn't be tightened (whereas spans like `/1/2` should).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235)*

**Comment by richardwu** [comment]
Review status: 0 of 20 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/aggregate, line 719 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-Kzk7mFAgM7nAkzA9WI5:-L-9ckqzgt1kbAnDKAWH:b6fnt0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/aggregate#L719)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'm thinking we can pass the direction info to prettyPrint. That might be helpful for #20203 too
</blockquote></details>

The pretty print with direction patch is in #20318.

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-KzkC8CdlwcJoh6Bgwqw:-L-9cayiXffQc6Hf6_Bp:bymy41t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L698)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, and also if the spans don't cover the entire primary index. If we have primary index on `a,b`, I believe we can't have interleaved child tables which share just `a` (right?). In this case, spans like `/1` shouldn't be tightened (whereas spans like `/1/2` should).
</blockquote></details>

Yup that's correct. Updated `TightenEndSpan` to reflect this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235#-KzkC8CdlwcJoh6Bgwqw:-L-CfvD_qVhjAncc1NMn:b-vyzqp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0605abb063708c710f847c2dd3f59c008999be91/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L698)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

Yup that's correct. Updated `TightenEndSpan` to reflect this.
</blockquote></details>

Cool, LGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20235)*

File: pkg/sql/opt_index_selection.go [hunk 1]
```diff
@@ -1153,15 +1167,21 @@ func spanFromLogicalSpan(
...
+
+			lastPartInclusive = false
 		}
 	}
-	s.EndKey = s.EndKey.PrefixEnd()
-	return s, nil
+
+	// We tighten the end key to prevent reading interleaved children
+	// after the last parent key.
+	s.EndKey, err = sqlbase.TightenEndKey(tableDesc, index, s.EndKey, lastPartInclusive)
```
Comment (RaduBerinde) [inline pkg/sql/opt_index_selection.go]: Sorry, I know this merged but I was working with this code and was confused for a while. Something non-obvious is going on here - we are not "tightening" the end key. If lastPartInclusive, we are actually "loosening" it to include the value pointing at `EndKey` (that's what `PrefixEnd()` used to do). Granted, we are loosening it as little as possible, but the name of the function is misleading. The comment here should also explain this.

Comment (RaduBerinde) [reply pkg/sql/opt_index_selection.go]: Ah, wait, no, we are doing `PrefixEnd` above, line 1147. I don't understand why the old code did `PrefixEnd()` here

Comment (RaduBerinde) [reply pkg/sql/opt_index_selection.go]: Ah, no, that's the start key, forget the second comment. The first one still stands.

Comment (richardwu) [reply pkg/sql/opt_index_selection.go]: Ah good point. Perhaps `AdjustEndKey` (and `AdjustStartKey`) with said comment(s)? It's less apparent what it does by the name but it's more "correct".

Comment (RaduBerinde) [reply pkg/sql/opt_index_selection.go]: Yeah, that works. Though I think `TightenStartKey` is accurate?
Also, I don't understand why the old code did `PrefixEnd` even in the `!lastPartInclusive` case?

Comment (richardwu) [reply pkg/sql/opt_index_selection.go]: If I understand your question correctly, the old code short-circuited if the last part is `!inclusive` here [at line 1160](https://github.com/cockroachdb/cockroach/pull/20235/files/048ec01a82dd5e01b9d0602b2cf575db2b68aef0#diff-e7c85a21eb65e3769380c3a0cbe91cc5L1160)

Comment (RaduBerinde) [reply pkg/sql/opt_index_selection.go]: Ah! Got it, thanks!

**Pull request merged**